pints/_abc/__init__.py

```diff
@@ -0,0 +1,362 @@
+        if logging:
+            if self._log_to_screen:
+                print('Using ' + str(self._sampler.name()))
+                if self._parallel:
+                    print('Running in parallel with ' + str(n_workers) +
+                          ' worker processess.')
+                else:
+                    print('Running in sequential mode.')
+
+            # Set up logger
+            logger = pints.Logger()
+            if not self._log_to_screen:
+                logger.set_stream(None)
+            if self._log_filename:
+                logger.set_filename(self._log_filename, csv=self._log_csv)
+
+            # Add fields to log
+            max_iter_guess = max(self._max_iterations or 0, 10000)
+            max_eval_guess = max_iter_guess
+            logger.add_counter('Iter.', max_value=max_iter_guess)
+            logger.add_counter('Eval.', max_value=max_eval_guess)
+            logger.add_float('Acceptance rate')
+            self._sampler._log_init(logger)
+            logger.add_time('Time m:s')
+
+        # Start sampling
+        timer = pints.Timer()
+        running = True
+
+        # Specifying the number of samples we want to get
+        # from the prior at once. It depends on whether we
+        # are using parallelisation and how many workers
+        # are being used.
+        if self._parallel:
+            n_requested_samples = 1
+        else:
+            n_requested_samples = self._n_workers
+
+        samples = []
+        # Sample until we find an acceptable sample
+        while running:
+            accepted_vals = None
+            while accepted_vals is None:
+                # Get points from prior
+                xs = self._sampler.ask(n_requested_samples)
+
+                # Simulate and get error
+                fxs = evaluator.evaluate(xs)
+                evaluations += self._n_workers
+
+                # Tell sampler errors and get list of acceptable parameters
+                accepted_vals = self._sampler.tell(fxs)
+
+            accepted_count += len(accepted_vals)
+            for val in accepted_vals:
+                samples.append(val)
+
+            iteration += 1
+
+            # Log progress
+            if logging and iteration >= next_message:
+                # Log state
+                logger.log(iteration, evaluations, (
+                           accepted_count / evaluations))
+                self._sampler._log_write(logger)
+                logger.log(timer.time())
+
+                # Choose next logging point
+                if iteration < self._message_warm_up:
+                    next_message = iteration + 1
+                else:
+                    next_message = self._message_interval * (
+                        1 + iteration // self._message_interval)
+
+            if iteration >= self._max_iterations:
+                running = False
+                halt_message = ('Halting: Maximum number of iterations ('
```

Comment:
In the case that we reach max_iterations before getting nr_samples, it would be nice for the halt message to warn the user. Something like "...Only {} samples were obtained!"